File: DSA/DynamicProgramming/Code/Memoization.cpp
#include <iostream>
#include <vector>
#include <cstring>
using namespace std;

// Memoization for Fibonacci
int fib(int n, vector<int>& dp) {
    if (n <= 1) return n;
    if (dp[n] != -1) return dp[n];
    return dp[n] = fib(n-1, dp) + fib(n-2, dp);
}

int main() {
    cout << "=== Memoization Example (C++) ===\n\n";
    int n = 10;
    vector<int> dp(100, -1);
    cout << "Fibonacci(" << n << ") = " << fib(n, dp) << endl;
    cout << "\nTime Complexity: O(n)\nSpace Complexity: O(n)\n";
    return 0;
}

File: DSA/DynamicProgramming/Code/Memoization.java
// Memoization Example in Java

public class Memoization {
    static int fib(int n, int[] dp) {
        if (n <= 1) return n;
        if (dp[n] != -1) return dp[n];
        return dp[n] = fib(n-1, dp) + fib(n-2, dp);
    }
    public static void main(String[] args) {
        System.out.println("=== Memoization Example (Java) ===\n");
        int n = 10;
        int[] dp = new int[100];
        for (int i = 0; i < 100; i++) dp[i] = -1;
        System.out.println("Fibonacci(" + n + ") = " + fib(n, dp));
        System.out.println("\nTime Complexity: O(n)");
        System.out.println("Space Complexity: O(n)");
    }
}
